File: src/backend/src/main/java/es/udc/hotelapp/backend/model/services/HotelService.java
package es.udc.hotelapp.backend.model.services;

import java.util.List;

import org.springframework.web.multipart.MultipartFile;

import es.udc.hotelapp.backend.model.entities.Hotel;
import es.udc.hotelapp.backend.model.entities.Product;
import es.udc.hotelapp.backend.model.entities.RoomTypePrice;
import es.udc.hotelapp.backend.model.entities.Service;
import es.udc.hotelapp.backend.model.exceptions.HotelAlreadyExistsException;
import es.udc.hotelapp.backend.model.exceptions.InstanceNotFoundException;
import es.udc.hotelapp.backend.model.exceptions.ProductAlreadyExistsException;
import es.udc.hotelapp.backend.model.exceptions.ServiceAlreadyExistsException;

public interface HotelService {

	/*************************** HOTEL **********************************/
	public Long createHotel(Hotel hotel) throws HotelAlreadyExistsException;

	public Hotel findById(Long id) throws InstanceNotFoundException;

	public Hotel updateHotel(Hotel hotel) throws InstanceNotFoundException;

	public void removeHotel(Long hotelid) throws InstanceNotFoundException;

	public Block<Hotel> findHotels(String name, String address, int page, int size);

	/*************************** PRODUCTS & SERVICES **********************************/
	public Long addService(Service service)
			throws InstanceNotFoundException, ServiceAlreadyExistsException;
	
	public Service findService(Long id) throws InstanceNotFoundException;
	
	public Block<Service> findServices(Long hotelid, String name, int page, int size);

	public Service updateService (Service service1) throws InstanceNotFoundException;
	
	public void removeService(Long serviceId) throws InstanceNotFoundException;

	public Long addProduct (Product p) throws InstanceNotFoundException, ProductAlreadyExistsException;

	public Product findProduct( Long id) throws InstanceNotFoundException;

	public Block<Product> findProducts(Long hotelid,String name, int page, int size);

	public Product updateProduct ( Product p) throws InstanceNotFoundException;

	public void removeProduct (Long productid) throws InstanceNotFoundException;

	/************************ ROOM PRICES  & PHOTOS **************************/
	
	public Long addPrice( RoomTypePrice price);
	
	public RoomTypePrice updateRoomTypePrice ( RoomTypePrice price) throws InstanceNotFoundException;
	
	public RoomTypePrice findPriceById( Long id);

	public void removePrice (Long priceid) throws InstanceNotFoundException;

	public boolean uploadPhoto ( MultipartFile file, Long hotelid);
	

}

File: src/frontend/src/backend/hotelService.js
import {config, appFetch} from './appFetch';


export const addHotel = (hotel, onSuccess, onErrors) =>
    appFetch(`/hotels`, config('POST', hotel), onSuccess, onErrors);


export const getHotels = (onSuccess) => {
    appFetch(`/hotels/`, config('GET'), onSuccess);
}

export const getHotelsById = (hotelid, onSuccess, onErrors) => {
    appFetch(`/hotels/${hotelid}`, config('GET'), onSuccess, onErrors);
}

export const updateHotel = (hotel, onSuccess, onErrors) =>{
    appFetch(`/hotels/${hotel.id}`, config('PUT', hotel), onSuccess, onErrors);
}

export const removeHotel = (hotel, onSuccess, onErrors) =>{
    appFetch(`/hotels/${hotel.id}`, config('DELETE'), onSuccess, onErrors);
}
/************************************************ PRODUCTS & SERVICES *********************************************** */
export const addService = (service, onSuccess, onErrors) =>{
    appFetch(`/hotels/${service.hotel.id}/services`, config('POST', service), onSuccess,onErrors);
}

export const addProduct = (product, onSuccess, onErrors) =>{
    appFetch(`/hotels/${product.hotel.id}/products`, config('POST', product), onSuccess,onErrors);
}

export const findServices = ({hotelid, page}, onSuccess, onErrors) =>{

    let path = `/hotels/${hotelid}/services/?page=${page}`;

    appFetch(path, config('GET'), onSuccess, onErrors);

}

export const findProducts = ({hotelid, page}, onSuccess, onErrors) =>{
    
    let path = `/hotels/${hotelid}/products/?page=${page}`;
    
    appFetch(path, config('GET'), onSuccess, onErrors);

}
export const findServiceById = (hotelid, serviceid, onSuccess, onErrors) => 
    appFetch(`/hotels/${hotelid}/services/${serviceid}`, config('GET'), onSuccess, onErrors);

export const findProductById = (hotelid, productid, onSuccess, onErrors) => 
    appFetch(`/hotels/${hotelid}/products/${productid}`, config('GET'), onSuccess, onErrors);

export const updateService = (service, onSuccess, onErrors) =>
    appFetch(`/hotels/${service.hotel.id}/services/${service.id}`, config('PUT', service), onSuccess, onErrors);

export const updateProduct = (product, onSuccess, onErrors) =>
    appFetch(`/hotels/${product.hotel.id}/products/${product.id}`, config('PUT', product), onSuccess, onErrors);

export const removeService = (hotelid, serviceid, onSuccess, onErrors) =>
    appFetch(`/hotels/${hotelid}/services/${serviceid}`, config('DELETE'), onSuccess, onErrors);

export const removeProduct = (hotelid, productid, onSuccess, onErrors) =>
    appFetch(`/hotels/${hotelid}/products/${productid}`, config('DELETE'), onSuccess, onErrors);

/**************************************************** PHOTOS **************************************************** */

export const uploadPhoto = (hotelid, file, onSuccess, onErrors) =>
    appFetch(`/hotels/${hotelid}/upload-photo`, config('POST', file), onSuccess, onErrors);

/************************************************ HOTEL PRICES ************************************************** */

export const addPrice = (price, onSuccess, onErrors) =>
    appFetch(`/hotels/${price.hotel.id}/prices`, config('POST', price), onSuccess, onErrors);

export const updatePrice = (price, onSuccess, onErrors) =>
    appFetch(`/hotels/${price.hotel.id}/prices/${price.id}`, config('PUT', price), onSuccess, onErrors);

export const findPriceById = (hotelid, priceid, onSuccess, onErrors) =>
    appFetch(`/hotels/${hotelid}/prices/${priceid}`, config('GET'), onSuccess, onErrors);

export const removePrice = (hotelid, priceid, onSuccess, onErrors) =>
    appFetch(`/hotels/${hotelid}/prices/${priceid}`, config('DELETE'), onSuccess, onErrors);



